src/bath.gleam

```diff
@@ -502,48 +724,120 @@ fn handle_pool_message(state: State(resource_type), msg: Msg(resource_type)) {
+            Ok(#(waiter, new_waiting)) -> {
               case state.create_resource() {
-                // Size hasn't changed
-                Ok(resource) -> #(
-                  deque.push_back(state.resources, resource),
-                  state.current_size,
-                )
-                // Size has changed
-                Error(resource_create_error) -> {
-                  log_resource_creation_error(
-                    state.log_errors,
-                    resource_create_error,
+                Ok(new_resource) -> {
+                  let selector = demonitor_process(selector, waiter.monitor)
+                  let #(monitor, selector) =
+                    monitor_process(selector, waiter.caller)
+                  let live_resources =
+                    dict.insert(
+                      live_resources,
+                      waiter.caller,
+                      LiveResource(resource: new_resource, monitor:),
+                    )
+                  actor.send(waiter.reply_to, Ok(new_resource))
+                  State(
+                    ..state,
+                    current_size: current_size + 1,
+                    live_resources:,
+                    waiting: new_waiting,
+                    selector:,
+                    resources: state.resources,
+                  )
+                  |> actor.continue
+                  |> actor.with_selector(selector)
+                }
+                Error(err) -> {
+                  log_resource_creation_error(state.log_errors, err)
+                  let selector = demonitor_process(selector, waiter.monitor)
+                  actor.send(
+                    waiter.reply_to,
+                    Error(CheckOutResourceCreateError(err)),
+                  )
+                  State(
+                    ..state,
+                    current_size:,
+                    live_resources:,
+                    waiting: new_waiting,
+                    selector:,
+                    resources: state.resources,
                   )
-                  #(state.resources, state.current_size)
+                  |> actor.continue
+                  |> actor.with_selector(selector)
                 }
               }
             }
-          }
+            Error(_) -> {
+              let #(new_resources, new_current_size) = case
+                state.creation_strategy
+              {
+                // If we create lazily, just decrement the current size - a new resource
+                // will be created when required
+                Lazy -> #(state.resources, current_size)
+                // Otherwise, create a new resource, warning if resource creation fails
+                Eager -> {
+                  case state.create_resource() {
+                    // Size hasn't changed
+                    Ok(resource) -> #(
+                      deque.push_back(state.resources, resource),
+                      current_size + 1,
+                    )
```

Comment:
I think the comments here are either wrong or the logic is

Reply:
Same as the other comment, this is net zero because 732 does -1 and this does +1. It's a bit confusing, so I want to refactor this to use some smaller helpers.